### PR TITLE
[TOPIC-BLE-LLCP] Bluetooth: Controller: Conn Update Procedure

### PR DESCRIPTION
**Pull request by tosk-ot**
A minimum of the Connection Update Procedure and its unit test is added.
The procedure does not yet contain all handling of actual functionality.
Unit tests have been prepared for PDU data and added support for
selected or unselected BT_CTLR_CONN_PARAM_REQ as well as its feature
bit.

Signed-off-by: Tommie Skriver <tosk@demant.com>

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_connections.c [hunk 1]
```diff
@@ -506,6 +516,16 @@ int ull_conn_llcp(struct ll_conn *conn, uint32_t ticks_at_expire, uint16_t lazy)
 {
 	LL_ASSERT(conn->lll.handle != ULL_HANDLE_NOT_CONNECTED);
 
+#if defined(CONFIG_BT_CTLR_CONN_PARAM_REQ)
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_connections.c:519]: I realized this is needed re. procedures that use instant (I had missed it re. phy update), so it should be passed in llcp and not llcp.conn_upd

**Review by erbr-ot — COMMENTED** [review]
I miss the update to ull_llcp_hci.c and respective unittest

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 2]
```diff
@@ -545,20 +549,63 @@ uint8_t ull_cp_conn_update(struct ll_conn *conn, uint16_t interval_min, uint16_t
...
+			uint16_t handle = ll_conn_handle_get(conn);
+
+			ull_slave_latency_cancel(conn, handle);
+		}
+		break;
+
+#endif /* CONFIG_BT_CTLR_CONN_PARAM_REQ */
+	default:
+		/* CPR feature not supported */
+		return BT_HCI_ERR_CMD_DISALLOWED;
```
[the remark this answers is not in the document]
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c:603]: I agree

File: tests/bluetooth/controller/common/src/helper_pdu.c [hunk 3]
```diff
@@ -1,557 +0,0 @@
-/*
```
Comment (thoh-ot) [inline tests/bluetooth/controller/common/src/helper_pdu.c]: Why is this file deleted in this commit?

File: subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h [hunk 4]
```diff
@@ -137,10 +137,36 @@ struct proc_ctx {
...
+			uint16_t offset0;
+			uint16_t offset1;
+			uint16_t offset2;
+			uint16_t offset3;
+			uint16_t offset4;
+			uint16_t offset5;
+			uint16_t *pdu_win_offset0;
+			uint32_t ticks_ref;
+			uint32_t ticks_to_offset_next;
+		} conn_param;
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:168]: Use shortform name `cpr`, so it matches the naming scheme of the rest of the structure

**Review by thoh-ot — CHANGES_REQUESTED** [review]
There seems to be a general data structure problem, `llcp_struct::conn_upd` and `llcp_struct::conn_param` er just legacy place holders that should be transformed into the correct place.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c:603]: I agree, too! Will replace with an assert.

File: tests/bluetooth/controller/ctrl_conn_update/CMakeLists.txt [hunk 5]
```diff
@@ -10,6 +10,10 @@ FILE(GLOB SOURCES
   src/*.c
 )
 
+if(CONFIG_BT_CTLR_CONN_PARAM_REQ)
+  add_compile_definitions(CONFIG_BT_CTLR_CONN_PARAM_REQ)
+endif()
+
```
[the remark this answers is not in the document]
Comment (tosk-ot) [reply tests/bluetooth/controller/ctrl_conn_update/CMakeLists.txt:16]: Yup. Will be removed.

File: tests/bluetooth/controller/ctrl_conn_update/testcase.yaml [hunk 6]
```diff
@@ -2,4 +2,10 @@ common:
     tags: test_framework bluetooth bt_conn_update bt_conn_param_req bt_ull_llcp
 tests:
   bluetooth.controller.ctrl_conn_update.test:
-     type: unit
+    type: unit
+    timeout: 5 # override 60s default timeout
```
[the remark this answers is not in the document]
Comment (tosk-ot) [reply tests/bluetooth/controller/ctrl_conn_update/testcase.yaml:6]: Yup. Will be removed.

File: tests/bluetooth/controller/ctrl_conn_update/testcase.yaml [hunk 7]
```diff
@@ -2,4 +2,10 @@ common:
...
+
+  bluetooth.controller.ctrl_conn_update.no_param_req_test:
+    type: unit
+    timeout: 5 # override 60s default timeout
```
[the remark this answers is not in the document]
Comment (tosk-ot) [reply tests/bluetooth/controller/ctrl_conn_update/testcase.yaml:10]: Yup. Will be removed.

File: tests/bluetooth/controller/common/src/helper_pdu.c [hunk 8]
```diff
@@ -0,0 +1,576 @@
+/*
```
[the remark this answers is not in the document]
Comment (tosk-ot) [reply tests/bluetooth/controller/common/src/helper_pdu.c]: This seems to be an unsuccessful attempt to move a file from one commit to another.
Will be fixed.

File: tests/bluetooth/controller/common/src/helper_pdu.c (hunk 3 above)
Comment (tosk-ot) [reply tests/bluetooth/controller/common/src/helper_pdu.c]: This seems to be an unsuccessful attempt to move a file from one commit to another.
Will be fixed.

File: subsys/bluetooth/controller/ll_sw/ull_connections.c (hunk 1 above)
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_connections.c:519]: Will move to llcp.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 9]
```diff
@@ -765,3 +1075,76 @@ void ull_cp_priv_rp_conn_param_req_neg_reply(struct ll_conn *conn, struct proc_c
...
+
+	LL_ASSERT(status == TICKER_STATUS_SUCCESS);
+
+	mfy.param = param;
+
+	/* Flush pending tx PDUs in LLL (using a mayfly) */
+	retval = mayfly_enqueue(TICKER_USER_ID_ULL_LOW, TICKER_USER_ID_LLL, 0,
+				&mfy);
+	LL_ASSERT(!retval);
+}
```
[the remark this answers is not in the document]
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:1150]: Yup. Will use that instead.

File: subsys/bluetooth/controller/ll_sw/ull_conn_types.h [hunk 10]
```diff
@@ -394,12 +394,13 @@ struct llcp_struct {
 		uint16_t instant;
 		uint16_t *pdu_win_offset;
 		uint32_t ticks_anchor;
+		uint32_t ticks_at_expire;	/* TODO: Place correct! */
+		uint16_t lazy;			/* TODO(tosk): Place correct! */
```
[the remark this answers is not in the document]
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_conn_types.h:398]: Will be moved.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 11]
```diff
@@ -167,14 +181,252 @@ static void lp_cu_ntf(struct ll_conn *conn, struct proc_ctx *ctx)
...
+		  (ticker_status == TICKER_STATUS_BUSY));
+
+#if (CONFIG_BT_CTLR_ULL_HIGH_PRIO == CONFIG_BT_CTLR_ULL_LOW_PRIO)
+	/* enable ticker job, if disabled in this function */
+	if (mayfly_was_enabled) {
+		mayfly_enable(TICKER_USER_ID_ULL_HIGH,
+			      TICKER_USER_ID_ULL_LOW, 1);
+	}
+#endif /* CONFIG_BT_CTLR_ULL_HIGH_PRIO == CONFIG_BT_CTLR_ULL_LOW_PRIO */
+}
```
[the remark this answers is not in the document]
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c]: Done.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h (hunk 4 above)
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h]: Done.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 12]
```diff
@@ -209,19 +234,31 @@ static void lp_cu_send_conn_update_ind(struct ll_conn *conn, struct proc_ctx *ct
 		ctx->state = LP_CU_STATE_WAIT_TX_CONN_UPDATE_IND;
 	} else {
 		/* TODO(thoh): Hardcoded instant delta +6 */
-		ctx->data.cu.instant = lp_event_counter(conn) + 6;
+		ctx->data.cu.instant = lp_event_counter(conn) + 6U;
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:237]: why not use the ull_conn_event_counter()? 
Also same comment as below re. define for delta

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 13]
```diff
@@ -495,12 +537,23 @@ static void rp_cu_conn_param_req_ntf(struct ll_conn *conn, struct proc_ctx *ctx)
 
 static void rp_cu_complete(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
 {
-	if (!ntf_alloc_is_available()) {
-		ctx->state = RP_CU_STATE_WAIT_NTF;
-	} else {
-		rp_cu_ntf(conn, ctx);
+	/*
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:540]: I'd prefer refactoring to not have complete-case code duplicated, but rather have a guard on the ntf to skip if not needed? Perhaps also change the prams_changed flag into a ntf flag - allowing the incorporation of the check for ctx->proc == PROC_CONN_UPDATE

**Review by erbr-ot — COMMENTED** [review]
WIP

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 14]
```diff
@@ -169,12 +182,23 @@ static void lp_cu_ntf(struct ll_conn *conn, struct proc_ctx *ctx)
...
-		ctx->state = LP_CU_STATE_WAIT_NTF;
-	} else {
-		lp_cu_ntf(conn, ctx);
+	/*
+	 * ConnUpd: Notify Host if any of the three conn. params have changed.
+	 * If no conn params are changed, the Host would not be notified.
+	 *
+	 * ParamReq: Notify Host.
+	 */
+	if ((ctx->proc == PROC_CONN_UPDATE) && (ctx->data.cu.params_changed == 0U)) {
```
Comment (tosk-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:191]: The (ctx->proc == PROC_CONN_UPDATE) && will be removed,
as shall always apply here.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c (hunk 13 above)
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:540]: As above, the check for ctx->proc == PROC_CONN_UPDATE does not apply and will be removed.
Agree with the duplicated code stuff - but this must come be later...

File: subsys/bluetooth/controller/ll_sw/ull_llcp_hci.c [hunk 15]
```diff
@@ -266,70 +268,21 @@ uint8_t ll_conn_update(uint16_t handle, uint8_t cmd, uint8_t status, uint16_t in
...
-			/*
-			 * EGON TODO: fill correct info in
-			 * the connection structure
-			 * as for example:
-			 *	conn->llcp_conn_param.status = status;
-			 *	conn->llcp_conn_param.state = cmd;
-			 *	conn->llcp_conn_param.cmd = 1U;
-			 */
+		if (status == 0U) {
+			ull_cp_conn_param_req_reply(conn);
```
Comment (tosk-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_hci.c:277]: I just realized that 'status' should be passed to ull_cp_conn_param_req_reply(), here:
Re.: 5.1.7.2: "If the Host rejects this request, then the device shall shall issue an LL_REJECT_EXT_IND PDU with the ErrorCode set to a value provided by the Host. The Host shall only use the error code Unacceptable Connection Parameters (0x3B) in order to reject the request."
So, it seems to be up to the Host to provide that error code...

File: subsys/bluetooth/controller/ll_sw/ull_conn_types.h [hunk 16]
```diff
@@ -391,34 +397,21 @@ struct llcp_struct {
 
 	/* EGON TODO: we'll need the next few structs eventually, Thomas and Szymon please comment on names etc. */
 	struct {
-		uint16_t instant;
 		uint16_t *pdu_win_offset;
 		uint32_t ticks_anchor;
 	} conn_upd;
 
+#if defined(CONFIG_BT_CTLR_LE_ENC)
```
Comment (tosk-ot) [inline subsys/bluetooth/controller/ll_sw/ull_conn_types.h:404]: Thus looks like must me messing up during resolving. Will be removed...

**Comment by tosk-ot** [comment]
Rebased and updated

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 17]
```diff
@@ -510,7 +563,7 @@ static void rp_cu_send_conn_update_ind(struct ll_conn *conn, struct proc_ctx *ct
 		ctx->state = RP_CU_STATE_WAIT_TX_CONN_UPDATE_IND;
 	} else {
 		/* TODO(thoh): Hardcoded instant delta +6 */
-		ctx->data.cu.instant = rp_event_counter(conn) + 6;
+		ctx->data.cu.instant = rp_event_counter(conn) + 6U;
```
[the remark this answers is not in the document]
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c]: Define made. Maybe we should make it common at one point, as you mentioned.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 18]
```diff
@@ -647,9 +766,24 @@ static void rp_cu_st_wait_rx_conn_update_ind(struct ll_conn *conn, struct proc_c
...
+		case BT_HCI_ROLE_MASTER:
+			ctx->unknown_response.type = PDU_DATA_LLCTRL_TYPE_CONN_UPDATE_IND;
+			rp_cu_send_unknown_rsp(conn, ctx, evt, param);
+			break;
+		case BT_HCI_ROLE_SLAVE:
+			pdu_decode_conn_update_ind(ctx, param);
+			//TODO(tosk): skip/terminate if instant passed?
+#if defined(CONFIG_BT_CTLR_CONN_PARAM_REQ)
+			/* conn param req procedure, if any, is complete */
+			conn->procedure_expire = 0U;
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:779]: void ull_conn_prt_clear(struct ll_conn *conn);

Should probably be changed to an inline, however it still does the thing;-)

**Review by erbr-ot — COMMENTED** [review]
Review in progress - have to continue tomorrow

File: subsys/bluetooth/controller/ll_sw/ull_connections.c [hunk 19]
```diff
@@ -223,7 +225,247 @@ void ll_rx_enqueue(struct node_rx_pdu *rx)
...
+
+	lll = &conn->lll;
+
+	/* Calculate current event counter */
+	event_counter = lll->event_counter + lll->latency_prepare + lazy;
+
+	return event_counter;
+}
+
+bool ull_conn_is_params_changed(struct ll_conn *conn, uint16_t interval,
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_connections.c:244]: Perhaps this should be in ull_llcp_conn_upd.c - so as to not 'load' ull_connections.c with cu specific code?
Perhaps we should have a discussion on this on Thursday?

File: subsys/bluetooth/controller/ll_sw/ull_connections.c [hunk 20]
```diff
@@ -223,7 +225,247 @@ void ll_rx_enqueue(struct node_rx_pdu *rx)
...
+
+	lll = &conn->lll;
+
+	/* Calculate current event counter */
+	event_counter = ull_conn_event_counter(conn);
+
+	instant_latency = (event_counter - instant) &
+			  0xFFFF;
+
+#if defined(CONFIG_BT_CTLR_CONN_PARAM_REQ)
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_connections.c:286]: I would put this outside of this function - and use the API as mentioned in other comment

Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_connections.c:286]: or place it next to the apto/sto handling below

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 21]
```diff
@@ -219,6 +222,7 @@ struct proc_ctx *ull_cp_priv_create_local_procedure(enum llcp_proc proc)
 	default:
 		/* Unknown procedure */
 		LL_ASSERT(0);
+		break;
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:225]: Is this break due to personal preference or coding guide lines?
Asking as we are not doing this elsewhere, so perhaps we should?

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 22]
```diff
@@ -183,8 +216,19 @@ static void lp_cu_send_conn_param_req(struct ll_conn *conn, struct proc_ctx *ctx
 	if (!tx_alloc_is_available() || rr_get_collision(conn)) {
 		ctx->state = LP_CU_STATE_WAIT_TX_CONN_PARAM_REQ;
 	} else {
+		uint16_t event_counter = lp_event_counter(conn);
+
 		rr_set_incompat(conn, INCOMPAT_RESOLVABLE);
 
+		ctx->data.cu.reference_conn_event_count = event_counter;
+		ctx->data.cu.preferred_periodicity = 0U;
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:224]: I'd personally prefer putting this initialization in a inline function or perhaps move it to init of the procedure.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 23]
```diff
@@ -209,19 +254,30 @@ static void lp_cu_send_conn_update_ind(struct ll_conn *conn, struct proc_ctx *ct
 		ctx->state = LP_CU_STATE_WAIT_TX_CONN_UPDATE_IND;
 	} else {
 		/* TODO(thoh): Hardcoded instant delta +6 */
-		ctx->data.cu.instant = lp_event_counter(conn) + 6;
+		ctx->data.cu.instant = lp_event_counter(conn) + CONN_UPDATE_INSTANT_DELTA;
+
 		lp_cu_tx(conn, ctx, PDU_DATA_LLCTRL_TYPE_CONN_UPDATE_IND);
-		ctx->rx_opcode = 0xFF; /* TODO(thoh): Hmm */
+		ctx->rx_opcode = PDU_DATA_LLCTRL_TYPE_UNUSED; /* TODO(thoh): Hmm */
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:260]: remove the 'hmm' comment?

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 24]
```diff
@@ -510,19 +595,19 @@ static void rp_cu_send_conn_update_ind(struct ll_conn *conn, struct proc_ctx *ct
 		ctx->state = RP_CU_STATE_WAIT_TX_CONN_UPDATE_IND;
 	} else {
 		/* TODO(thoh): Hardcoded instant delta +6 */
-		ctx->data.cu.instant = rp_event_counter(conn) + 6;
+		ctx->data.cu.instant = rp_event_counter(conn) + CONN_UPDATE_INSTANT_DELTA;
 		rp_cu_tx(conn, ctx, PDU_DATA_LLCTRL_TYPE_CONN_UPDATE_IND);
-		ctx->rx_opcode = 0xFF; /* TODO(thoh): Hmm */
+		ctx->rx_opcode = PDU_DATA_LLCTRL_TYPE_UNUSED; /* TODO(thoh): Hmm */
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:600]: 'hmm'

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 25]
```diff
@@ -566,7 +672,19 @@ static void rp_cu_st_wait_rx_conn_param_req(struct ll_conn *conn, struct proc_ct
 {
 	switch (evt) {
 	case RP_CU_EVT_CONN_PARAM_REQ:
-		rp_cu_send_conn_param_req_ntf(conn, ctx, evt, param);
+		pdu_decode_conn_param_req(ctx, param);
+
+		bool parameters_is_changed = ull_conn_is_params_changed(conn,
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:677]: Nit-pick (capital N) spelling: parameters_ARE_changed or better jet just loose the 'is' (also in function!)

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 21 above)
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c:225]: Misra.
https://gitlab.com/MISRA/MISRA-C/MISRA-C-2012/Example-Suite/-/blob/master/R_16_03.c

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c (hunk 18 above)
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:779]: Done

File: subsys/bluetooth/controller/ll_sw/ull_connections.c (hunk 19 above)
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_connections.c:244]: Agree! Moved (and renamed)

File: subsys/bluetooth/controller/ll_sw/ull_connections.c (hunk 20 above)
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_connections.c:286]: I'm not sure I know what you mean?

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c (hunk 23 above)
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:260]: Yes.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c (hunk 24 above)
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:600]: Yes.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c (hunk 25 above)
Comment (tosk-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:677]: Sure! Function and parameters renamed!

**Comment by tosk-ot** [comment]
Rebased, updated and fixed compliance errors.

File: subsys/bluetooth/controller/ll_sw/ull_connections.c (hunk 20 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_connections.c:271]: No, I see - looks like the comment was put on the wrong line

Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_connections.c:271]: I'm thinking of the handling of conn->procedure_expire, I'd move that code to where the other timeout related functionality is in this function (re. apto and sto)

Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_connections.c:271]: But its a nit pick!

**Review by erbr-ot — APPROVED** [review]
LGTM

**Review by thoh-ot — APPROVED** [review]
